src/topotoolbox/stream_object.py

```diff
@@ -0,0 +1,211 @@
+                    cell_area = self.cellsize**2
+                else:
+                    # False so cellsize is in degrees
+                    pass
+        else:
+            err = (f"Invalid unit '{units}' provided. Expected one of "
+                   f"'pixels', 'mapunits', 'm2', 'km2'.")
+            raise ValueError(err) from None
+
+        # Create the appropriate threshold matrix based on the threshold input.
+        if isinstance(threshold, (int, float)):
+            if threshold == 0:
+                avg = (flow.shape[0] + flow.shape[1])//2
+                threshold = np.full(
+                    self.shape, math.floor((avg**2)*0.01), dtype=np.float32)
+            else:
+                threshold = np.full(self.shape, threshold, dtype=np.float32)
+        elif isinstance(threshold, np.ndarray):
+            if threshold.shape != self.shape:
+                err = (f"Threshold array shape {threshold.shape} does not "
+                       f"match FlowObject shape: {self.shape}.")
+                raise ValueError(err) from None
+            threshold = threshold.astype(np.float32, order='F')
+        else:
+            if threshold.shape != self.shape:
+                err = (f"Threshold GridObject shape {threshold.shape} does "
+                       f"not match FlowObject shape: {self.shape}.")
+                raise ValueError(err) from None
+
+            threshold = threshold.z.astype(np.float32, order='F')
+
+        # Divide the threshold by how many m^2 or km^2 are in a cell to convert
+        # the user input to pixels for further computation.
+        threshold /= cell_area
+
+        # Generate the flow accumulation matrix (acc)
+        acc = np.zeros_like(flow.target, order='F', dtype=np.float32)
+        weights = np.ones_like(flow.target, order='F', dtype=np.float32)
+        _flow.flow_accumulation(
+            acc, flow.source, flow.direction, weights, flow.shape)
+
+        # Generate a 1D array that holds all indexes where more water than
+        # in the required threshold is collected. (acc >= threshold)
+        threshold = threshold.flatten(order='F')
+        acc = acc.flatten(order='F')
+        temp = []
+        for i, value in enumerate(acc):
+            if value >= threshold[i]:
+                temp.append(i)
+
+        self.stream = np.array(temp, dtype=np.int32)
+
+        # Based on the stream array, generate 3 1D arrays where the value of
+        # the stream array at each index holds respective value of the
+        # original array. (source, target and direction)
+        source = flow.source.flatten(order='F')
+        temp_source = []
+        target = flow.target.flatten(order='F')
+        temp_target = []
+        direction = flow.direction.flatten(order='F')
+        temp_direction = []
+
+        for i in self.stream:
+            temp_source.append(source[i])
+            temp_target.append(target[i])
+            temp_direction.append(direction[i])
```

Comment:
Same thing here as above: building up these lists one by one is likely to be slow. But you can index with the `stream` array:

```python
source = flow.source.ravel(order='F')[stream]
```
etc.